txjuju/testing/api.py

```diff
@@ -195,20 +197,48 @@ def _formatAnnotationInfo(self, info, verb):
 
     def _formatApplicationInfo(self, info, verb):
         if self.version.startswith("2."):
-            return ["application", verb, {
+            formatted = {
                 "name": info.name,
-                "charm-url": info.charmURL}]
+                "exposed": info.exposed,
+                }
+            if info.charmURL is not None:
+                formatted["charm-url"] = info.charmURL
```

Comment:
Since we positively expect all those fields to be there (and IIRC the current Go code in Juju will include them all under all circumstances, but I didn't check again now), my preferred solution would be to:

1) Have the various formatters in FakeAPIBackend always include all the fields that we expect a real backend would include (so essentially removing the various "if" checks in this branch).

2) Modify Juju2APIClient._parseAllWatcherNextDelta slightly to log a warning if a field is not there. For example:

```python
    if kind == "unit":
            status, statusInfo = self._getDeltaJujuStatus(data)
            info = UnitInfo(
                data[self._getParam("name")],
                data[self._getParam("application")],
                series=data.get(self._getParam("series")),
```

could become something like:

```python
    if kind == "unit":
            status, statusInfo = self._getDeltaJujuStatus(data)
            info = UnitInfo(
                self._getParam(data, "name"),
                self._getParam("application"),
                series=self._getParam(data, "series"),
```

Where _getParam logs the warning in case the given field is missing (this just an example, modify at will with any semantically equivalent factoring you might prefer).

From my memory, the historical reason for those "data.get" in _parseAllWatcherNextDelta is that back then we didn't have a convinent way to fabricate deltas (like FakeAPIBackend), so the "get" part was meant to be test-friendly (you didn't have to set data all the times).

Perhaps now we can get past this small "hack" and be stricter (although still have a graceful fallback with a warning, for future proof against Juju API server changes).